helix-core/src/syntax.rs

```diff
@@ -293,9 +302,25 @@ impl Loader {
                     .and_then(|extension| self.language_config_ids_by_file_type.get(extension))
             });
 
-        configuration_id.and_then(|&id| self.language_configs.get(id).cloned())
+        // If we have not found the configuration_id, see if we can get it from a shebang line
+        if configuration_id.is_none() {
+            if let Ok(mut file) = File::open(path) {
+                let mut buf = [0; 100];
+                if std::io::Read::read(&mut file, &mut buf[..]).is_ok() {
+                    if let Ok(str) = str::from_utf8(&buf) {
+                        static SHEBANG_REGEX: Lazy<Regex> = Lazy::new(|| {
+                            Regex::new(r"#!/[^\s]*/(env\s)*([_a-zA-Z0-9-]+)").unwrap()
```

Comment:
I took a pass at cleaning up the shebang regex. Here's an alternative to yours that captures only the interpreter part, and allows more exotic interpreter names (like "php7.4"). 

```regex
^#!\s*/\S*/(?:env\s+)?(\S+)
```

I wrote a bunch of tests over here so you can see how it breaks down and modify if needed:
https://regex101.com/r/fsxGMO/2

Reply:
Thanks for the regex cleanup!
I made some minor changes to also support lines like "#! python"
One thing im not sure of, if we really want to match for minor versions of interpreters, is there a real world usecase for this?
This makes it harder to cut off the .exe/.cmd/.. part of interpreters, while having the need o support minor versions in language toml shebangs entry.

what i have now (excluding minor versions):
```
^#!\s*(?:\S*/(?:env\s+)?)?([^\s\.]+)
```

Reply:
@ath3 good point about supporting things like `php7.4`. My thought was that sometimes you might have more than one version of php installed on a server (like with ondrej's ubuntu repos) that name them this way, and you _may possibly_ want to specifically invoke one. But for our case, we just want to discover the name of the language, not the version, so you're modification makes sense. Although, it does still erroneously catch `php7` out of my example. Do we want to exclude digits from the capture group? That would also clean up the `python3` case.

Also, for windows, would we need to worry about supporting `\` as a directory separator? If so, just modify the `/` in your version to `[/\\]`.

Reply:
> Also, for windows, would we need to worry about supporting `\` as a directory separator?

I don't think windows supports shebangs, not natively anyway ([cygwin can use](https://superuser.com/questions/484041/run-shell-script-on-windows-using-cygwin) it I think, with `/` like in unix).

Reply:
I  searched and saw that some people use Windows style paths in shebang lines which is probably not the most correct way, but we can use a regex handling those cases too.

Reply:
I pushed the new regex, hopefully this works for all the variants.